### PR TITLE
Implement a wrapper that opens new shards

### PR DESCRIPTION
**Pull request by kralka**
Implement DatasetFiller which automatically takes care of opening new
shards with proper part and group number. Refactor the code to use it.

Fixes issue #58

### CODE REVIEW

File: scaaml/io/dataset.py [hunk 1]
```diff
@@ -276,8 +279,9 @@ def new_shard(self, key: list, part: int, group: int, split: str,
         if split not in ["train", "test", "holdout"]:
             raise ValueError("Invalid split, must be: {train, test, holdout}")
 
-        if part < 0 or part > 10:
-            raise ValueError("Invalid part value -- muse be in [0, 10]")
+        if part < 0 or part > self.MAX_PART_NUMBER:
+            msg = "Invalid part value -- muse be in [0, self.MAX_PART_NUMBER]"
```
Comment (jmichelp) [inline scaaml/io/dataset.py:283]: typo: must

Also, why creating a temp variable here?

File: scaaml/io/dataset_filler.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      examples_per_shard (int): Number of examples in a shard.
+      plaintexts_per_key (int): Number of plaintexts used with a concrete key.
+      repetitions (int): Number of examples belonging to the same key,
+        plaintext pair.
+      skip_examples (int): How many examples to skip (non-negative integer
+        divisible by examples_per_shard). Defaults to zero.
+    """
+
+    def __init__(self,
+                 dataset,
```
Comment (jmichelp) [inline scaaml/io/dataset_filler.py:36]: is the no typing here intended?

File: scaaml/io/dataset_filler.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                attack_points=attack_points,
+                measurement=measurement,
+                current_key=current_key,
+                split_name=split_name,
+                chip_id=chip_id,
+            )
+    """
+
+    def __init__(self,
+                 dataset,
```
Comment (jmichelp) [inline scaaml/io/dataset_filler.py:162]: Same as above regarding typing.

File: tests/io/test_dataset_filler.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    """Test with."""
+    mock_dataset = MagicMock()
+    mock_dataset.examples_per_shard = 64
+
+    with DatasetFiller(
+            dataset=mock_dataset,
+            plaintexts_per_key=256,
+            repetitions=1,
+    ) as dataset_filler:
+        # Assure that the context type is right.
```
Comment (jmichelp) [inline tests/io/test_dataset_filler.py:35]: typo? Ensure?

File: scaaml/io/__init__.py [hunk 5]
```diff
@@ -14,3 +14,4 @@
 """Dataset creation and usage."""
 
 from .dataset import Dataset  # noqa
+from .dataset_filler import DatasetFiller  # noqa
```
[the remark this answers is not in the document]
Comment (kralka) [reply scaaml/io/__init__.py]: Fixed this file, created an issue https://github.com/google/scaaml/issues/82 (not fixing everything in this PR).

File: scaaml/io/dataset.py (hunk 1 above)
Comment (kralka) [reply scaaml/io/dataset.py]: Seemed like a nicer formatting.

File: scaaml/io/dataset.py [hunk 6]
```diff
@@ -289,8 +291,9 @@ def new_shard(self, key: list, part: int, group: int, split: str,
         if split not in Dataset.SPLITS:
             raise ValueError(f"Invalid split, must be in {Dataset.SPLITS}")
 
-        if part < 0 or part > 10:
-            raise ValueError("Invalid part value -- muse be in [0, 10]")
+        if part < 0 or part > self.MAX_PART_NUMBER:
+            raise ValueError(
+                "Invalid part value -- must be in [0, self.MAX_PART_NUMBER]")
```
Comment (jmichelp) [inline scaaml/io/dataset.py:296]: Shouldn't that be an f-string with `self.MAX_PART_NUMBER` surrounded by curly braces?

**Review by jmichelp — APPROVED** [review]

**Pull request merged**